### PR TITLE
Add max user uses to discounts

### PR DESCRIPTION
**Pull request by ryanmitchell**
This PR adds the ability to limit the number of times a user can apply a discount code. 

It adds a new relation between discounts and users with a pivot table
It tracks when a logged in user uses a discount
It adds a new field to the admin hub to "Max uses per user", when this has a value if there is no logged in user then it will fail to add the discount.

### CODE REVIEW

**Comment by glennjacobs** [comment]
@ryanmitchell can you avoid using `Auth::user()` and instead use `$cart->user` ?

File: packages/core/database/migrations/2023_03_03_100001_add_max_uses_per_user_to_discounts_table.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+use Lunar\Base\Migration;
+
+class AddMaxUsesPerUserToDiscountsTable extends Migration
+{
+    public function up()
+    {
+        Schema::table($this->prefix.'discounts', function (Blueprint $table) {
+            $table->json('max_uses_per_user')->nullable()->after('max_uses');
```
Comment (alecritson) [inline packages/core/database/migrations/2023_03_03_100001_add_max_uses_per_user_to_discounts_table.php:12]: Should this be a json field?

File: packages/core/src/DiscountTypes/AbstractDiscountType.php [hunk 2]
```diff
@@ -39,6 +41,10 @@ public function markAsUsed(): self
     {
         $this->discount->uses = $this->discount->uses + 1;
 
+        if ($user = Auth::user()) {
```
[the remark this answers is not in the document]
Comment (ryanmitchell) [reply packages/core/src/DiscountTypes/AbstractDiscountType.php]: Thanks - should be updated now

File: packages/core/database/migrations/2023_03_03_100001_add_max_uses_per_user_to_discounts_table.php (hunk 1 above)
Comment (ryanmitchell) [reply packages/core/database/migrations/2023_03_03_100001_add_max_uses_per_user_to_discounts_table.php]: Definitely not.

**Comment by alecritson** [comment]
@ryanmitchell Is this all finished up and ready?

**Comment by ryanmitchell** [comment]
Yeah it is

**Pull request merged**